src/AI/JakeMan.java

```diff
@@ -0,0 +1,770 @@
+            int damage = Math.min(defender.getHP(), (int)results.defender.getPreciseHPDamage());
+
+            // Convert to abstract value
+            int extraLoss = 0;
+            if( loss >= 1 && unit.getHP() == UnitModel.MAXIMUM_HP )
+              extraLoss += STAY_UNHURT_BIAS;
+            if( loss >= unit.getHP() )
+              extraLoss += STAY_ALIVE_BIAS;
+            loss *= unit.getCost();
+            loss += extraLoss;
+            damage *= defender.getCost();
+            if( isThreatenedBy(unit.model, defender.model) )
+              damage *= FIRSTSTRIKE_ON_THREAT_WEIGHT;
+            // Value damage to hurt units less
+            damage *= defender.getHPFactor();
+
+            return (double)(damage - loss);
+          }, (terrain, params) -> 0.01); // Attack terrain, but don't prioritize it over units
+
+      if( damageValue > bestDamage )
+      {
+        bestDamage = damageValue;
+        bestAttack = attack;
+      }
+    }
+    return bestAttack;
+  }
+
+  private static boolean isThreatenedBy(UnitModel um, UnitModel threat)
+  {
+    int threshhold = um.hasDirectFireWeapon() ? DIRECT_THREAT_THRESHHOLD : INDIRECT_THREAT_THRESHHOLD;
+    boolean isThreat = false;
+    for( WeaponModel wm : threat.weapons )
+      isThreat |= threshhold <= wm.getDamage(um);
+    return isThreat;
+  }
+
+  private static boolean isWeakTo(UnitModel um, UnitModel threat)
+  {
+    boolean isWeak = isThreatenedBy(um, threat);
+    isWeak &= !isThreatenedBy(threat, um);
+    return isWeak;
+  }
+
+  private boolean isDudeFree(GameMap gameMap, Unit unit, XYCoord xyc, boolean amAttacking)
+  {
+    HashMap<UnitModel, Double> threatCounts = new HashMap<>();
+    for( UnitModel threat : unitMapEnemy.keySet() )
+    {
+      if( !isThreatenedBy(unit.model, threat) )
+        continue;
+      if( unitMapEnemy.get(threat).containsKey(xyc) )
+        threatCounts.put(threat, unitMapEnemy.get(threat).get(xyc));
+    }
+    if( threatCounts.size() < 1 )
+      return true;
+
+    ArrayList<XYCoord> counterCoords = Utils.findLocationsInRange(gameMap, xyc, 1);
+    counterCoords.remove(xyc);
+    for( UnitModel threat : threatCounts.keySet().toArray(new UnitModel[0]) )
+      for( UnitModel counter : unitMapFriendly.keySet() )
+      {
+        if( !isThreatenedBy(threat, counter) )
+          continue;
+        double counterPowerTotal = 0;
+        final boolean counterIsMeAndIAmPeaceful = !amAttacking && (counter == unit.model);
+        for( XYCoord coord : counterCoords )
+        {
+          double counterPower = 0;
+          if( unitMapFriendly.get(counter).containsKey(coord) )
+            counterPower = unitMapFriendly.get(counter).get(coord);
+          if( counterIsMeAndIAmPeaceful )
+            counterPower -= PEACEFUL_SELF_THREAT_RATIO * unit.getHPFactor();
+          counterPowerTotal += Math.max(0, counterPower);
+        }
+        final double counterPowerAverage = counterPowerTotal / counterCoords.size();
```

Comment:
Huh. Taking the average is an interesting choice. If you have a 10HP unit and a 1HP unit of the same type, it would look much worse than if you just had the 10HP unit.

Reply:
That takes the average of the *tiles*, not of the *units*.
Units are additive (should maybe cap to the top N units, where N is the number of available spaces? IDK), but each unit may not threaten all tiles that an enemy could attack me from... so I take the average.
Maybe it should instead only count units that can threaten all tiles a direct could attack me from?

Reply:
> That takes the average of the tiles, not of the units.

Ah. Still an interesting choice. A 10HP unit with a single attack square is much more apparently dangerous than a 10HP unit with four attack options.
A "correct" solution might be to take into account the specific counter locations reachable by each countering unit and allocate slots, then add up the most potent permutation. But eh.

Reply:
> A 10HP unit with a single attack square is much more apparently dangerous than a 10HP unit with four attack options.

Still no, unless I am really misremembering/misreading. The sums of the threats on each square are averaged.
So 2 tanks threatening one square is considered equivalent to 1 tank threatening 2 squares.
The above is kind of a fundamental problem with this method, but I don't see a way to resolve it without getting a lot more complicated.